x/shield/keeper/withdraw.go

```diff
@@ -105,3 +130,218 @@ func (k Keeper) DequeueCompletedWithdrawQueue(ctx sdk.Context) {
 	k.SetTotalCollateral(ctx, totalCollateral)
 	k.SetTotalWithdrawing(ctx, totalWithdrawing)
 }
+
+// ComputeWithdrawAmountByTime computes the amount of collaterals
+// from a given provider that will be dequeued from the withdraw
+// queue by a given time.
+func (k Keeper) ComputeWithdrawAmountByTime(ctx sdk.Context, provider sdk.AccAddress, time time.Time) sdk.Int {
+	withdrawTimesliceIterator := k.WithdrawQueueIterator(ctx, time)
+	defer withdrawTimesliceIterator.Close()
+
+	amount := sdk.ZeroInt()
+	for ; withdrawTimesliceIterator.Valid(); withdrawTimesliceIterator.Next() {
+		timeslice := []types.Withdraw{}
+		value := withdrawTimesliceIterator.Value()
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(value, &timeslice)
+
+		for _, withdraw := range timeslice {
+			if withdraw.Address.Equals(provider) {
+				amount = amount.Add(withdraw.Amount)
+			}
+		}
+	}
+	return amount
+}
+
+func (k Keeper) ComputeTotalUnbondingAmount(ctx sdk.Context, provider sdk.AccAddress) sdk.Int {
+	unbondings := k.sk.GetAllUnbondingDelegations(ctx, provider)
+
+	sum := sdk.ZeroInt()
+	for _, ubd := range unbondings {
+		for i := 0; i < len(ubd.Entries); i++ {
+			sum = sum.Add(ubd.Entries[i].Balance)
+		}
+	}
+	return sum
+}
+
+func (k Keeper) ComputeUnbondingAmountByTime(ctx sdk.Context, provider sdk.AccAddress, time time.Time) sdk.Int {
+	dvPairs := k.getUnbondingsByProviderMaturingByTime(ctx, provider, time)
+
+	sum := sdk.ZeroInt()
+	seen := make([]sdk.ValAddress, 0, len(dvPairs))
```

Comment:
Why not use a map?

Reply:
I think because Golang maps are not deterministic.

Reply:
But why the non-determinism of map here will be a problem?

Reply:
I think using a map here is fine.

Reply:
Right. Fixed.